### ISSUE TITLE
Allow an Application to prevent login when user is not registered

### ISSUE THREAD

**Issue by robotdan** [FusionAuth/fusionauth-issues#439]
### Problem
The current intended design of FusionAuth is to authenticate the user and then allow the application to determine if they user is authorized to the application. 

Authorization is done by including the `applicationId` and `roles` claims in the JWT and on the OAuth2 Authorization Code Grant redirect we also include a `userState` parameter that indicates if the user is `Authenticated` or `AuthenticatedNotRegistered`. 

There are scenarios where the application would prefer that we verify a base level of authorization by ensuring the user is registered for the application represented by the `client_id` in an OAuth2 grant, or the `applicationId` on the Login API. 

### Solution
Add a toggle in the FusionAuth Application to require user registration to get a successful login response. 

This means that unless a user is registered for the application represented by the `client_id` in an OAuth grant, or the `applicationId` on the Login API the request will fail. We could return a status code of `403` for example instead of `401` which is returned when the credentials are invalid. 

### Alternatives/workarounds
The application can look at the `userState` value or rely upon the JWT for authorization since this is the primary mechanism used to verify the user authorization. 

### Additional context
N/A

### How to vote
Please give us a thumbs up or thumbs down as a reaction to help us prioritize this feature. Feel free to comment if you have a particular need or comment on how this feature should work.

**Comment by sascha-schieferdecker** [comment]
I think this a very good idea. I just implemented a TYPO3 extension for registering with FusionAuth. IMHO the registration process would be a lot smoother to handle clientwise if FusionAuth would return a 403 status code because the user should not get access because the verification of the email adress is not completed yet.

**Comment by mreschke** [comment]
If it were our own application using FusionAuth then your idea of letting our code "determine if they user is authorized" is fine...but with third party apps this is not always possible.  I found this thread while integrating ERPNext to use FusionAuth as its 'social login' feature using the oauth2 client and secret.  Everything works as expected except for this registration problem.  FA user has zero registrations yet ERPNext logs in just fine.  FA knows the client_id from URL, it should double check the user has that app (as an optional setting as you suggest).

**Comment by davidmw** [comment]
I'd like to see this go a step further. If Self Registration is off, but Google SSO is enabled - anyone with a valid Google hosted email account can attempt a login, resulting in the creation of an account on FusionAuth with no app registrations.

We manage user accounts from our app. A new user gets an account on our backend and is created and registered on FusionAuth via the API.

Today we created a user account with a typo in the email address. User was instructed to login. They connected with valid Google SSO account, which FusionAuth couldn't match up to the existing account so created a new one. That new FusionAuth account wasn't authorized to do any thing so user complained.

So our admin spotted the typo in the email address and attempted to fix in our FE. FusionAuth rejects the email address change since an account with the email already and newly exists on the system.

Our workaround, I suppose, is to detect the email change request when the user record is edited, delete the FusionAuth account via API, and recreate the account rather than update it on FusionAuth. Just seems rather unnecessary if the Google ghost account wasn't created in the first place.

**Comment by davidmw** [comment]
Actually the workaround described above is risky - not a great idea to delete an account causing an email address collision since it might be a valid account. Might be safer if we first confirm that target account has no registrations.

**Comment by bostjan** [comment]
@robotdan, am I correct to assume that SAML-based assertions are equally affected by this same behavior?

@mreschke hits the nail on the head in the comment above - you can't control what third party apps support wrt SSO.

Therefore, issuing valid SAML assertions without authorization data (roles/groups/whatever) and only relying on the other side to validate the authorization data before allowing access can potentially be considered a security issue. At least SAML-based SPs often have something that resembles a "default access role".

**My (lousy) workaround**
The only way I can make the FusionAuth _not_ issue a valid SAML Assertion (for user that exists, is authenticated, but has no registration for this application and this application's self registration is disabled) is by creating a lambda that checks the `registration` object and throws an error if it equals `null`:
```javascript
function populate(samlResponse, user, registration) {
  if (registration === null) {
    throw 'User ' + user.email + ' is not authorized (registered) to use this application (acs=' + samlResponse.destination + ', audience[0]=' + samlResponse.assertion.conditions.audiences[0] + '), and this application does not allow the user to self-register.';
  }
}
```

This workaround has a few caveats:
- The message only appears in the Event Log, the user only sees a generic error message.
- Only actual registrations and applications' self-registration setting influences the result.
(This workaround is probably more related to #952 and/or #969.)

**Suggestion**
Personally, IMHO, I'd prefer to see this solved in connection with the Groups feature:
- Add a third option for self-registration switch, in addition to `yes` and `no` - `Allow self-registration ONLY when user is in one of the groups that have access (=defined role) for a given application`
- Decide what to do when the group's definition for this application changes/is removed:
  - Does such registration stay in FusionAuth indefinitely, or
  - Does such registration keep a reference to the Group's role definition, and when that reference changes (i.e. the group is removed from the application, or another role is selected for this group), the related registrations are deleted too (to enforce the new authorization)?

**Comment by robotdan** [comment]
> and only relying on the other side to validate the authorization data before allowing access can potentially be considered a security issue.

I would have to agree, but this is a security issue on the SP end, not the IdP. If any SP accepts a third party JWT, SAML assertion and does not validate it, assert the subject, audience, and authorization claims etc - that is a bug in the SP.

**Comment by bostjan** [comment]
@robotdan, I think I understand where you're coming from ("[Built for Developers](https://fusionauth.io/features/built-for-developers/)"), and from a technical standpoint, I fully agree with you.

However, most likely the reality is that each organization has:
- One (or a few) application they fully control (i.e. the main product) and are therefore able to adjust its behavior to the chosen IdP.
- Many external apps (either self-hosted or SaaS) to create a functional work environment (i.e. Gmail, Slack, Jira, SalesForce, Github, CI, CD for devs and ops, then there are marketing tools, analytics, legal teams have their own stuff I don't even pretend to know, and HR and accounting and so on), over which they have very little or no control wrt how SP part is implemented.

All I am trying to say is it seems counter-productive to artificially make your product harder (or impossible) to use in cases where your customers can't fully control the SP side, but they do need that SP to perform their business function. Once you make your customers/prospects try out other IdP alternatives, you're risking losing them to those alternatives.

PS: Don't get me wrong, I am not a potential FusionAuth customer at this point, I am just using it for personal use. Thanks for sharing this great product with the world for free, and I don't expect you to jump on anything I say. But I do see the needs of the company that I work for, and I think what we're discussing here would be a potential show stopper if FusionAuth would be evaluated as a replacement IdP. Therefore, the reason why I am sharing my (humble!) opinion here is to contribute to FusionAuth's success (in a very minor! way) or to help it not needlessly miss out on opportunities, that's all. Happy Easter! ;)

**Comment by voidmain** [comment]
@bostjan - just to add to what @robotdan was saying, we want to be careful but labeling this as a "security issue" within FusionAuth.

Currently, FusionAuth is correctly validating the user's identity (email and password for example). When using OIDC, we return an extra parameter called `UserState` that indicates if the user is registered for the application or not. For SAML, we were including that same parameter in previous versions, but it didn't work for some SPs, so we removed it.

Even still, I'm not positive this is a "security issue". Here's why my thinking leans this way. When you have an SP that is protected (i.e. Gmail, Slack, Jira, SalesForce, Github, CI, CD, etc), they are not going to just let any user in just because they successfully authenticated with the IdP. All of these SPs will require that the user already exist in the SP database and will match them up based on some unique identifier, usually an email address. Therefore, it doesn't matter what the response from FusionAuth is.

If you have unprotected SPs (apps that allow everyone in and users to register themselves), then you would enable "self registration" in FusionAuth. Once this is enabled, the user would always be registered to the application as part of the authentication process. The SP would then setup everything for the user in its own database. This scenario doesn't require any blocking for unregistered users.

These two scenarios are the main ones I've seen in the real-world. If I'm missing other real-wold scenarios that could potentially be security risks, please let me know.

With all of that said, it is on our roadmap to add a way to prevent the login if the user isn't registered for the application. My view is that this is a feature rather than a "security issue" though.

**Comment by bostjan** [comment]
@voidmain, thanks for (re:)explaining.

About the "security issue" - to soften the impact of my words, I used the word "potentially" next to it in my comment above. But I guess that didn't work as well as I intended. I didn't mean to do any fear mongering, sorry if it came across like that.

**Comment by robotdan** [comment]
Available in 1.28.0, doc is still in progress. But you will find a new configuration in the Application OAuth2 configuration.
